### PR TITLE
[feat] Deposit reward when adding a new template

### LINKED ISSUE #30: [feat] Template ERC20 rewards deposit
Add the functionality to deposit the `reward` from the `reward_token` in the `add_template` function call. Make sure to revert if the deposit fails at any step.

#### References
- [Template add_template func](https://github.com/keep-starknet-strange/art-peace/blob/5d71f79a5fc1a7fbfd758bc2646d9e26c48f5cfb/onchain/src/templates/component.cairo#L55)
- [ERC20 deposit example](https://github.com/keep-starknet-strange/succinct-starknet/blob/11b2139427a34b5aaf764eab4db77942e8b5fb60/src/fee_vault.cairo#L144)

### PR DESCRIPTION
**Pull request by TAdev0**
- [x] issue #30 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/art-peace/blob/main/CONTRIBUTING.md)
- [x] code change includes tests
- [x] breaking change

Closes #30 

@b-j-roberts  i had issues with conflicts, i opened a new PR.

I added tests for depositing reward.
Tell me if this is ok for you

### CODE REVIEW

**Comment by TAdev0** [comment]
@b-j-roberts  ready for review

File: onchain/src/templates/component.cairo [hunk 1]
```diff
@@ -59,11 +62,38 @@ pub mod TemplateStoreComponent {
             let template_id = self.templates_count.read();
             self.templates.write(template_id, template_metadata);
             self.templates_count.write(template_id + 1);
+            self
```
Comment (b-j-roberts) [inline onchain/src/templates/component.cairo:65]: Lets have this logic run only if `reward_token` and `reward` are non-zero, IE lets allow templates to be added with no reward.

File: onchain/src/templates/component.cairo [hunk 2]
```diff
@@ -59,11 +62,38 @@ pub mod TemplateStoreComponent {
...
         }
     }
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState, +HasComponent<TContractState>
+    > of InternalTrait<TContractState> {
+        fn deposit(
+            ref self: ComponentState<TContractState>,
+            template_proposer: ContractAddress,
```
Comment (b-j-roberts) [inline onchain/src/templates/component.cairo:83]: `template_proposer` arg can be removed and just use the `caller_address` like you did in the function.

**Review by b-j-roberts — CHANGES_REQUESTED** [review]
Couple small things.

**Comment by b-j-roberts** [comment]
General tip: You should be able to do a force push after resolving conflicts, which would prevent you from needing to open a new PR. Something like `git push ... -f`

**Comment by TAdev0** [comment]
@b-j-roberts yeah I know , I just had issues when rebasing. Thanks !

File: onchain/src/templates/component.cairo (hunk 2 above)
Comment (TAdev0) [reply onchain/src/templates/component.cairo:83]: oh yeah true, my bad. thanks!

File: onchain/src/templates/component.cairo (hunk 1 above)
Comment (TAdev0) [reply onchain/src/templates/component.cairo:65]: you mean IE lets NOT allow templates to be added with no reward right?

Comment (b-j-roberts) [reply onchain/src/templates/component.cairo:65]: I want to allow templates without rewards. Currently, if there is no reward token, the deposit function will revert with `Invalid token`.  So, `deposit` should only be called when `reward` and `reward_token` are set to non-zero values.

**Comment by TAdev0** [comment]
@b-j-roberts  PR updated !

**Comment by b-j-roberts** [comment]
Nice work! Thanks for quickly resolving issues.

**Comment by TAdev0** [comment]
@b-j-roberts  if you have any other issue, i'm available :)

**Pull request merged**